### PR TITLE
Filtered queries

### LINKED ISSUE #107: Adapt backend and APIs to filter queries
Adapt the backend that was implemented via #69 and APIs that were implemented via #76 to filtered queries.

### PR DESCRIPTION
**Pull request by CatoLeanTruetschel**
This PR implements query capabilities such that a generic filter can be specified.

### CODE REVIEW

**Review by joluj — COMMENTED** [review]
I will review that in detail later or even tomorrow.
Here are only some first comments.

File: backend/src/filter/FilterConditionBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+} from '../shared/queries';
+import { allocateParamKey } from './allocateParamKey';
+import FilterConditionVisitor from './FilterConditionVisitor';
+
+export interface QueryParams {
+  [key: string]: unknown | undefined;
+}
+
+export default class FilterConditionBuilder extends FilterConditionVisitor {
+  /* eslint-disable lines-between-class-members */
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/FilterConditionBuilder.ts:16]: Tracked by #146

File: backend/src/filter/FilterConditionBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  public constructor(entity: 'node' | 'edge', name: string) {
+    super();
+    this.entity = entity;
+    this.name = name;
+  }
+
+  public buildCondition(filter: FilterCondition): [string | null, QueryParams] {
+    this.visit(filter);
+    return [this.resultCondition, this.queryParams];
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/FilterConditionBuilder.ts:31]: Tracked by #147

File: backend/src/filter/FilterConditionVisitor.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (condition.rule === 'any') {
+      // TODO: Check whether required properties are set.
+      return this.visitMatchAnyCondition(<MatchAnyCondition>condition);
+    }
+
+    return condition;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/FilterConditionVisitor.ts:36]: Tracked by #142

File: backend/src/filter/allocateParamKey.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { QueryParams } from './FilterConditionBuilder';
+
+export function allocateParamKey(params: QueryParams, name: string): string {
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/allocateParamKey.ts:3]: Tracked by  #141

File: backend/src/filter/FilterConditionBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  OfTypeCondition,
+} from '../shared/queries';
+import { allocateParamKey } from './allocateParamKey';
+import FilterConditionVisitor from './FilterConditionVisitor';
+
+export interface QueryParams {
+  [key: string]: unknown | undefined;
+}
+
+export default class FilterConditionBuilder extends FilterConditionVisitor {
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/FilterConditionBuilder.ts:15]: Tracked by  #141

File: backend/src/filter/filter.service.ts [hunk 6]
```diff
@@ -1,16 +1,99 @@
...
+
+    const result = await this.neo4jService.read(query, params);
+    return result.records.map((x) => x.toObject() as NodeDescriptor);
+  }
+
+  private async getEdges(
+    filter?: FilterCondition,
+    limit?: number
+  ): Promise<EdgeDescriptor[]> {
+    const [condition, params] = buildFilterCondition('edge', 'e', filter);
```
[the remark this answers is not in the document]
Comment (CatoLeanTruetschel) [reply backend/src/filter/filter.service.ts:77]: This is the some thing as [here](https://github.com/amosproj/amos-ss2021-project2-context-map/pull/115#discussion_r632926646).
Maybe also just add some comments? Tracked by #141

**Pull request merged**